### ISSUE TITLE
Gracefully handle exceptions in command handlers

### ISSUE THREAD

**Issue by alexhelms** [dotnet/command-line-api#1655]
When an exception occurs in a command handler the entire stack trace is barfed out into stderr and I'd like to not have that behavior for explicitly handled exceptions. Consider a handler that tries to read a file but the filename is not found. The standard libraries throw a FileNotFoundException which I'd like to catch and display just the message and not the entire stack trace. For unhandled exceptions I think the current behavior of the stack trace is fine since by definition it is unhandled.

I looks like I can get the invocation context and set the invocation result to a custom IInvocationResult. The problem is there doesn't appear to be (or I missed it 😅) an ErrorResult for setting a user-defined error that the handler failed. I made my own that is essentially a copy of `ParseErrorResult` but you can specify the message in the ctor.

```c#
public class ErrorResult : IInvocationResult
{
    private readonly string _errorMessage;
    private readonly int _errorExitCode;

    public ErrorResult(string errorMessage, int errorExitCode = 1)
    {
        _errorMessage = errorMessage;
        _errorExitCode = errorExitCode;
    }

    public void Apply(InvocationContext context)
    {
        context.Console.ResetTerminalForegroundColor();
        context.Console.SetTerminalForegroundRed();

        context.Console.Error.WriteLine(_errorMessage);
        context.Console.Error.WriteLine();

        context.ExitCode = _errorExitCode;

        context.Console.ResetTerminalForegroundColor();
    }
}
```

The main issue I ran into was that `ResetTerminalForegroundColor` and `ResetTerminalForegroundColor` are internal extension methods and those also rely on the internal `Platform` static class. I copied them into my code base for now.

Assuming those APIs were public, ErrorResult now works and you can use it like in the following trivial example:

```c#
void Handler(InvocationContext ctx, string filename)
{
    try
    {
        Console.WriteLine(File.ReadAllText(filename));
    }
    catch (FileNotFoundException e)
    {
        ctx.InvocationResult = new ErrorResult(e.Message, 42);
    }
}
```

Now when the handled exception occurs, it doesnt bubble up and kill the application but instead just prints (in red) to stderr. 

I'm not familiar with the details of how System.CommandLine works. Do you think this is a valid approach for handling exceptions in a handler?

**Comment by jonsequitur** [comment]
You can customize the exception handling behavior a bit more directly by using the `CommandLineBuilder.UseExceptionHandler` method: 

https://github.com/dotnet/command-line-api/blob/d86f87aa98d6392f99a407930e2b6451f6a33372/src/System.CommandLine/Builder/CommandLineBuilderExtensions.cs#L291-L301

**Comment by alexhelms** [comment]
Thanks, I'll have a look. Do you think there is value in having a user defined `IInvocationResult` like I showed?

**Comment by jonsequitur** [comment]
My suggestion was for an approach that might be simpler for the specific goal you mentioned but a custom `IInvocationResult` will give you more control if the scenario is more complex. I think it really depends on how much of the default behavior you want to replace.

**Comment by alexhelms** [comment]
I agree that the general exception handler will accomplish what I want. What I mean is do you think having user defined `IInvocationResult` is the way to go for situations where you want to differ the error situation when the general handler isn't enough or would become too complex? Hypothetically, I use the general exception handler to print just the exception message but if I wanted some commands to have different behavior what would be the best way to achieve that?

**Comment by jonsequitur** [comment]
I think it's largely a matter of how you prefer to structure your app and what kinds of outputs you need to produce. 

If I have an error condition that's specific to a command and is something you might expect in the context of that one command command, I prefer to handle that in a `try...catch` within the handler. It's easier to understand the code and it's easier to test in the absence of the command line parser infrastructure.

On the other hand, if I want to enforce a common exception handling behavior across all commands, middleware and a custom `IInvocationResult` is a good choice.
